src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala

```diff
@@ -219,13 +227,154 @@ object SnowflakeTelemetry {
       metric.put(TelemetryOOBFields.EXCEPTION_STACKTRACE, stringWriter.toString)
     }
     metric.put(TelemetryQueryStatusFields.DETAILS, details)
+    if (config.isDefined) {
+      metric.set(TelemetryQueryStatusFields.SYSTEM_CONFIG, config.get)
+    }
 
     SnowflakeTelemetry.addLog(
       (TelemetryTypes.SPARK_QUERY_STATUS, metric),
       System.currentTimeMillis()
     )
     SnowflakeTelemetry.send(conn.getTelemetry)
   }
+
+  // Configuration retrieving is optional for for diagnostic purpose,
+  // so it never raises exception.
+  private[snowflake] def getSystemConfigWithoutTaskInfo(): ObjectNode = {
+    val metric: ObjectNode = mapper.createObjectNode()
+
+    // System config retrieving is for diagnostic purpose, it raises exception.
+    try {
+      addSystemConfigInternal(metric)
+    } catch {
+      case _: Throwable => {
+        metric
+      }
+    }
+  }
+
+  private def addSystemConfigInternal(metric: ObjectNode): ObjectNode = {
+    // Add versions info
+    Utils.addClientInfoJson(metric)
+    // Add JVM system config
+    addJVMConfig(metric)
+    // Add Spark option
+    addSparkConfig(metric)
+  }
+
+  // Configuration retrieving is optional for for diagnostic purpose,
+  // so it never raises exception.
+  private[snowflake] def getSystemConfigWithTaskInfo(): ObjectNode = {
+    val metric = getSystemConfigWithoutTaskInfo()
+
+    try {
+      addSystemConfigInternal(metric)
+      // Add task and executor info
+      addTaskInfo(metric)
+    } catch {
+      case _: Throwable => {
+        metric
+      }
+    }
+  }
+
+  private def addJVMConfig(metric: ObjectNode): ObjectNode = {
+    // System basic information
+    metric.put(TelemetryConfigFields.OS_NAME,
+      System.getProperty(TelemetryConstValues.JVM_PROPERTY_NAME_OS_NAME))
+    val rt = Runtime.getRuntime
+    metric.put(TelemetryConfigFields.MAX_MEMORY_IN_MB, rt.maxMemory() / MB)
+    metric.put(TelemetryConfigFields.TOTAL_MEMORY_IN_MB, rt.totalMemory() / MB)
+    metric.put(TelemetryConfigFields.FREE_MEMORY_IN_MB, rt.freeMemory() / MB)
+    metric.put(TelemetryConfigFields.CPU_CORES, rt.availableProcessors())
+    metric
+  }
+
+  // Below specifies the spark options may affects spark connector
```

Comment:
To reviewers:
Actually, we could collect all spark options. But, I spend time to go through spark options to identify the options which may affect SC. The reason is that I find some spark options are credential. It may be a problem if they are logged or sent to telemetry.
The disadvantage is that some options may be missed.

Please advice any others spark option needs to be collected. Or any options are not good to collect.

Reply:
Does it make more sense to have a blacklist instead of a whitelist here?

Reply:
The difficulty is how to build a complete blacklist.
Before building the whitelist, I considered to skip spark options whose name includes "key"/"password"/"credential" (ignore case).
But I think whitelist is safer, so I give up that proposal.